src/pid/environ.rs

```diff
@@ -0,0 +1,179 @@
+}
+
+impl<'a> EnvironIter<'a> {
+    fn parse(&self) -> Option<Result<(&'a [u8], (OsString, OsString))>> {
+        match pair(self.data) {
+            IResult::Done(data, parsed) => Some(Ok((data, parsed))),
+            IResult::Incomplete(_) => None,
+            IResult::Error(err) => Some(Err(Error::new(
+                ErrorKind::InvalidInput,
+                format!("Unable to parse input: {:?}", err),
+            ))),
+        }
+    }
+}
+
+impl<'a> Iterator for EnvironIter<'a> {
+    type Item = Result<(OsString, OsString)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.data.len() == 0 {
+            return None;
+        }
+        let (data, item) = match self.parse() {
+            None => return None,
+            Some(Err(x)) => return Some(Err(x)),
+            Some(Ok(x)) => x,
+        };
+        self.data = data;
+        Some(Ok(item))
+    }
+}
+
+impl<'a> IntoIterator for &'a Environ {
+    type Item = Result<(OsString, OsString)>;
+    type IntoIter = EnvironIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        EnvironIter {
+            data: &self.data,
+        }
+    }
+}
+
+/// Extracts name of a variable. Also consumes a delimiter.
+fn get_name(src: &[u8]) -> IResult<&[u8], OsString> {
+    // Calculate position of the equal sign.
+    let pos = match src.iter().skip(1).position(|c| c == &b'=') {
+        Some(p) => p,
+        None => return IResult::Error(error_position!(nom::ErrorKind::Custom(0), src)),
+    };
+    IResult::Done(&src[pos + 2..], from_bytes(&src[..pos + 1]))
+}
+
+#[test]
+fn test_get_name() {
+    let src = &b"FOO=BAR"[..];
+    assert_eq!(get_name(src), IResult::Done(&b"BAR"[..], "FOO".into()));
+    let src = &b"FOO="[..];
+    assert_eq!(get_name(src), IResult::Done(&b""[..], "FOO".into()));
+    let src = &b"=FOO=BAR"[..];
+    assert_eq!(get_name(src), IResult::Done(&b"BAR"[..], "=FOO".into()));
+    let src = &b"=FOO="[..];
+    assert_eq!(get_name(src), IResult::Done(&b""[..], "=FOO".into()));
+}
+
+/// Executes `take_until_and_consume` and maps its result to OsString.
+named_args!(
+    until<'a>(sep: &'a str)<OsString>,
+    map!(take_until_and_consume!(sep), from_bytes)
+);
+
+/// Parses 'key=value" pair.
+named!(
+    pair<&[u8], (OsString, OsString)>,
+    tuple!(get_name, apply!(until, "\0"))
+);
+
+#[test]
+fn test_pair() {
+    let source = &b"FOO=BAR\0123"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"123"[..], ("FOO".into(), "BAR".into()))
+    );
+    let source = &b"FOO=\0123"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"123"[..], ("FOO".into(), "".into()))
+    );
+    let source = &b"=FOO=BAR\0-"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"-"[..], ("=FOO".into(), "BAR".into()))
+    );
+    let source = &b"=FOO=\0-"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"-"[..], ("=FOO".into(), "".into()))
+    );
+}
+
+
+#[test]
+fn test_iter() {
```

Comment:
Wrap all of the tests in a `#[cfg(test)] mod tests {` like other files, and put them all at the end of the file.